# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/gamepad/GamepadHandle.cpp
# Commit: ec6a64b639a8
# Full Hash: ec6a64b639a82e957376c09a71aef7ce62b3a5aa
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2020-12-03 09:47:26
# Regressor Bug: 1657404
# File Overlap Count: 1
# Description:
#   Bug 1657404 - Implement a strongly-typed, service-dependent gamepad handle r=handyman,aklotz
#   
#   Currently, the gamepad code uses a uint32_t as a handle and does some trickery
#   with it by trying to create a unique ID and adding an offset to it for VR code.
#   
# ==============================================================================

diff -r 6792f28b99bd -r ec6a64b639a8 dom/gamepad/GamepadHandle.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/gamepad/GamepadHandle.cpp	Wed Dec 02 23:06:05 2020 +0000
@@ -0,0 +1,40 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "GamepadHandle.h"
+#include "mozilla/Assertions.h"
+#include "mozilla/HashFunctions.h"
+
+namespace mozilla::dom {
+
+GamepadHandle::GamepadHandle(uint32_t aValue, GamepadHandleKind aKind)
+    : mValue(aValue), mKind(aKind) {
+  MOZ_RELEASE_ASSERT(mValue);
+}
+
+GamepadHandleKind GamepadHandle::GetKind() const { return mKind; }
+
+PLDHashNumber GamepadHandle::Hash() const {
+  return HashGeneric(mValue, uint8_t(mKind));
+}
+
+bool operator==(const GamepadHandle& a, const GamepadHandle& b) {
+  return (a.mValue == b.mValue) && (a.mKind == b.mKind);
+}
+
+bool operator!=(const GamepadHandle& a, const GamepadHandle& b) {
+  return !(a == b);
+}
+
+bool operator<(const GamepadHandle& a, const GamepadHandle& b) {
+  if (a.mKind == b.mKind) {
+    return a.mValue < b.mValue;
+  }
+  // Arbitrarily order them by kind
+  return uint8_t(a.mKind) < uint8_t(b.mKind);
+}
+
+}  // namespace mozilla::dom
\ No newline at end of file